recirq/otoc/parallel_xeb.py

```diff
@@ -0,0 +1,725 @@
+            FSIM unitaries, represented by cirq.Circuit objects, as values.
+        correction_gates: Same as fitted_gates, but with all Z rotations
+            reversed in signs.
+        fitted_angles: A dictionary with qubit pairs as keys and optimized
+            FSIM unitaries as values. Here the FSIM unitaries are represented
+            as a dictionaries with the names of the FSIM phases as keys and
+            their fitted values as values.
+        final_errors_optimized: A dictionary with qubit pairs as keys and
+            their cycle errors after fitting as values.
+        final_errors_unoptimized: A dictionary with qubit pairs as keys and
+            their cycle errors before fitting as values.
+    """
+    num_trials = len(measured_bits[0])
+    p_data_all, sq_gates = _pairwise_xeb_probabilities(
+        all_qubits, num_cycle_range, measured_bits, scrambling_gates,
+        interaction_sequence)
+    final_errors_unoptimized = {}
+    final_errors_optimized = {}
+    delta_angles = {}
+    purity_errors = {}
+    fitted_gates = {}
+    fitted_angles = {}
+    correction_gates = {}
+
+    for (q0, q1), p_data in p_data_all.items():
+        print('Fitting qubits {} and {}'.format(q0, q1))
+
+        def xeb_fidelity(angle_shifts: np.ndarray, num_p: int
+                         ) -> Tuple[float, List[float], np.ndarray,
+                                    np.ndarray, float]:
+            new_angles = fsim_angles.copy()
+            for i, angle_name in enumerate(_fsim_angle_labels):
+                new_angles[angle_name] += angle_shifts[i]
+            fsim_mat = angles_to_fsim(**new_angles)
+
+            max_cycles = num_cycle_range[num_p - 1]
+
+            p_sim = [np.zeros((num_trials, 4)) for _ in
+                     range(max_cycles)]
+            for i in range(num_trials):
+                unitary = np.identity(4, dtype=complex)
+                for j in range(max_cycles):
+                    mat_0 = _rot_mats[sq_gates[(q0, q1)][i][0, j]]
+                    mat_1 = _rot_mats[sq_gates[(q0, q1)][i][1, j]]
+                    unitary = np.kron(mat_0, mat_1).dot(unitary)
+                    unitary = fsim_mat.dot(unitary)
+                    if j + 1 in num_cycle_range:
+                        idx = num_cycle_range.index(j + 1)
+                        p_sim[idx][i, :] = np.abs(unitary[:, 0]) ** 2
+
+            fidelities = [_alpha_least_square(p_sim[i], p_data[i]) for i in
+                          range(num_p)]
+
+            cost = -np.sum(fidelities)
+
+            err, x_vals, y_vals = pauli_error_fit(
+                np.asarray(num_cycle_range)[0:num_p],
+                np.asarray(fidelities), add_offset=False)
+
+            return err, fidelities, x_vals, y_vals, cost
+
+        def cost_function(angle_shifts: np.ndarray) -> float:
+            if gate_to_fit == 'sqrt-iswap':
+                full_shifts = np.zeros(5, dtype=float)
+                full_shifts[1:4] = angle_shifts
```

Comment:
lots of magic indices in this function. Add comments explaining or (better yet) use something more self-documenting like a dictionary

Reply:
Comments added to explain these indices.